File: debugging/tests/fixtures/A3/charge.py
class GatewayError(Exception):
    pass


class _PaymentGateway:
    def charge(self, amount):
        raise GatewayError("gateway timeout")   # the gateway can fail


payment_gateway = _PaymentGateway()


class Order:
    def __init__(self, total):
        self.total = total
        self.paid = False

    def mark_paid(self, txn_id):
        self.paid = True
        self.txn_id = txn_id


def charge(order):
    resp = payment_gateway.charge(order.total)   # can raise GatewayError
    order.mark_paid(resp.id)

File: debug/tests/fixtures/A3/charge.ts
export class GatewayError extends Error {}

export const paymentGateway = {
  charge(total: number): { id: string } {
    if (total > 100) {
      throw new GatewayError("declined");
    }
    return { id: "tx_" + total };
  },
};

export interface Order {
  total: number;
  paidId: string | null;
  failedReason: string | null;
  markPaid(id: string): void;
  markFailed(reason: string): void;
}

export function charge(order: Order): void {
  const resp = paymentGateway.charge(order.total); // can throw GatewayError
  order.markPaid(resp.id);
}
